src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/TagHelperDescriptorJsonConverter.cs

```diff
@@ -906,5 +914,141 @@ private static void ReadToEndOfCurrentObject(JsonReader reader)
 
             throw new JsonSerializationException($"Could not read till end of object, end of stream. Got '{reader.TokenType}'.");
         }
+
+        private static string Intern(string str)
+        {
+            if (str is null)
+            {
+                return null;
+            }
+
+            var interned = string.IsInterned(str);
+            if (interned != null)
+            {
+                return interned;
+            }
+
+            return weakReferenceTable.Intern(str);
+        }
+
+        /// <summary>
+        /// The purpose of this class is to avoid permanently storing duplicate strings which were read in from JSON.
+        /// </summary>
+        private class WeakReferenceTable
+        {
+            public static readonly WeakReferenceTable Instance = new WeakReferenceTable();
+
+            private WeakReferenceTable()
+            {
+            }
+
+            private readonly WeakValueDictionary<string, string> _table = new WeakValueDictionary<string, string>();
+
+            public string Intern(string str)
+            {
+                if (_table.TryGetValue(str, out var value))
+                {
+                    return value;
+                }
+                else
+                {
+                    _table.Add(str, str);
+                    return str;
+                }
+            }
+
+            // ConditionalWeakTable won't work for us because it only compares to Object.ReferenceEquals.
+            private class WeakValueDictionary<TKey, TValue> where TValue : class
+            {
+                // Your basic project.razor.json using our template have ~800 entries,
+                // so lets start the dictionary out large
+                private int _capacity = 400;
+
+                // This might not acctually need to be concurrent?
```

Comment:
There's other threading concerns in this class I'd enforce "multi-threadedness" at the consumer layer and not worry about it here